src/Fixer/ArrayNotation/ReturnToYieldFromFixer.php

```diff
@@ -0,0 +1,82 @@
+<?php
+
+declare(strict_types=1);
+
+/*
+ * This file is part of PHP CS Fixer.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *     Dariusz Rumiński <dariusz.ruminski@gmail.com>
+ *
+ * This source file is subject to the MIT license that is bundled
+ * with this source code in the file LICENSE.
+ */
+
+namespace PhpCsFixer\Fixer\ArrayNotation;
+
+use PhpCsFixer\AbstractFixer;
+use PhpCsFixer\FixerDefinition\CodeSample;
+use PhpCsFixer\FixerDefinition\FixerDefinition;
+use PhpCsFixer\FixerDefinition\FixerDefinitionInterface;
+use PhpCsFixer\Tokenizer\Analyzer\FunctionsAnalyzer;
+use PhpCsFixer\Tokenizer\CT;
+use PhpCsFixer\Tokenizer\Token;
+use PhpCsFixer\Tokenizer\Tokens;
+
+/**
+ * @author Kuba Werłos <werlos@gmail.com>
+ */
+final class ReturnToYieldFromFixer extends AbstractFixer
+{
+    public function getDefinition(): FixerDefinitionInterface
+    {
+        return new FixerDefinition(
+            'When function return type is iterable and it starts with `return` then it must be changed to `yield from`.',
+            [new CodeSample('<?php function giveMeData(): iterable {
+    return [1, 2, 3];
+}
+')],
+        );
+    }
+
+    public function isCandidate(Tokens $tokens): bool
+    {
+        return $tokens->isAnyTokenKindsFound([T_ARRAY, CT::T_ARRAY_SQUARE_BRACE_OPEN]);
+    }
+
+    /**
+     * {@inheritdoc}
+     *
+     * Must run before YieldFromArrayToYieldsFixer.
+     * Must run after PhpUnitDataProviderReturnTypeFixer, PhpdocToReturnTypeFixer.
+     */
+    public function getPriority(): int
+    {
+        return 1;
+    }
+
+    protected function applyFix(\SplFileInfo $file, Tokens $tokens): void
+    {
+        $functionsAnalyzer = new FunctionsAnalyzer();
+
+        foreach ($tokens->findGivenKind(T_FUNCTION) as $index => $token) {
+            $returnType = $functionsAnalyzer->getFunctionReturnType($tokens, $index);
+            if (null === $returnType || 'iterable' !== strtolower($returnType->getName())) {
```

Comment:
```suggestion
            if (null === $returnType || 'array' === strtolower($returnType->getName())) {
```
+ `?array`, ˙array|null˙, `array|false` etc.

anything with object implies `Traverseable` which should be fixed.

Reply:
Take a look at https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/pull/7168/commits/d8d9cddf862bdbd81bd7b06542d9957797b780d2 - that would fail with this change (with the function returning `int`).

We would have to check if after the `return` there is `array(`, or `[`, or `\array_merge`, or `array_merge` (without `use BetterArray\array_merge` above), or `array_filter`, or... I think I made a point 😉

Reply:
yes, with return `array(` / `[` :)

Reply:
I don't want to limit the fixer to only those 2 cases.

Reply:
my point is how `iterable` is so special vs. `Travesable` (`iterable~array`)

Reply:
I thought that `iterable` is only one legal because it can have `return [/* ... */]` and `yield from`, but other build-in types are fine too when we think of `return getGenerator();`.

Take a look at [latest commit](https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/pull/7168/commits/8593a18fb4a8388e9d9fccde2e6526182f6b5b53), WDYT?

Reply:
I have to apologise, converting `return $this->getIterator()` to `yield from $this->getIterator()` does not make any sense. It would imply additional yield->iterator step instead of just returning the original iterator.

You were originally right the only sensual return type is `iterable`, as the fixer makes sense for `return [` only. For `array` return type, we cannot return iterator (as the result would not be an array anymore) nor an array can be returned when the type is `Iterator` - https://3v4l.org/Tb8Y0.

Reply:
Yeah, I realized that when I applied the extended version to this repo code, it changed [this](https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/blob/v3.22.0/src/Console/ConfigurationResolver.php#L638) and broke tests :)